src/transformers/trainer.py

```diff
@@ -360,14 +366,24 @@ def __init__(
         data_collator: Optional[DataCollator] = None,
         train_dataset: Optional[Union[Dataset, IterableDataset]] = None,
         eval_dataset: Optional[Union[Dataset, Dict[str, Dataset]]] = None,
-        tokenizer: Optional[PreTrainedTokenizerBase] = None,
+        processor: Optional[
+            Union[PreTrainedTokenizerBase, BaseImageProcessor, SequenceFeatureExtractor, ProcessorMixin]
+        ] = None,
         model_init: Optional[Callable[[], PreTrainedModel]] = None,
         compute_metrics: Optional[Callable[[EvalPrediction], Dict]] = None,
         callbacks: Optional[List[TrainerCallback]] = None,
         optimizers: Tuple[torch.optim.Optimizer, torch.optim.lr_scheduler.LambdaLR] = (None, None),
         preprocess_logits_for_metrics: Optional[Callable[[torch.Tensor, torch.Tensor], torch.Tensor]] = None,
-        image_processor: Optional["BaseImageProcessor"] = None,
+        tokenizer: Optional[PreTrainedTokenizerBase] = None,
     ):
+        if tokenizer is not None:
+            warnings.warn(
+                "The `tokenizer` argument is deprecated and will be removed in v5 of Transformers. You can use `processor` "
```

Comment:
I believe this PR is proposing deprecating `tokenizer` @amyeroberts! I think `preprocessor` is limited in the sense that we don't just want to save the pre-processing class (e.g. image processor or feature extractor), but also the post-processing class (e.g. the tokenizer) => for this reason I think passing the `processor` is cleanest here (for multimodal models)

Reply:
Ah, yes. Perhaps I wasn't clear. What I meant was we would introduce an new argument, which should be used in preference to `tokenizer` which would accept all of the processing class. However we wouldn't fully deprecate `tokenizer` for a long time, and still allow it to be used

Reply:
Great thanks for the clarification - on the same page here! Happy to update the PR #30864 accordingly, unless you want to see this one to completion @NielsRogge?